Adding micro-ROS with Teensy and Arduino IDE tutorial (#325)
micro-ROS/micro-ROS.github.io: _docs/tutorials/core/micro_ROS_with_teensy_arduino/index.md

---

> title: Micro-ROS with Teensy and Arduino IDE

Author: ```suggestion
title: Teensy with Arduino
```

> title: Micro-ROS with Teensy and Arduino IDE

Author: ```suggestion
title: Teensy with Arduino
```

> permalink: /docs/tutorials/core/micro_ROS_with_teensy_arduino/

Author: ```suggestion
permalink: /docs/tutorials/core/Teensy_with_Arduino/
```

> permalink: /docs/tutorials/core/micro_ROS_with_teensy_arduino/

Author: Changed below

micro-ROS client library for our demonstration we will not be going to build the firmware and 
thus we will skip the firmware build steps from the same [*page*](https://micro.ros.org//docs/tutorials/core/first_application_linux/#:~:text=Installing%20ROS%202%20and%20the%20micro-ROS%20build%20system).

To install the micro-ros-agent follow the steps below:

```bash
# Download micro-ROS-Agent packages
ros2 run micro_ros_setup create_agent_ws.sh
```
We will now build the agent packages and, when this is done, source the installation:

```bash
# Build step
ros2 run micro_ros_setup build_agent.sh
source install/local_setup.bash
```
Now, let's give a dry run by running the micro-ros-agent by following the command:

```bash
ros2 run micro_ros_agent micro_ros_agent serial --dev /dev/ttyACM0
```
The result should show something like this:

![Illustration of agent running](micro_ros_agent_start.png)

This means the installation of the micro-ros-agent is successful. 
Now we can proceed to the next step which is the installation of Arduino IDE 
and Teensyduino and patching the Arduino-based Teensy board for using the 

> pre-compiled libraries as described [*here*](https://github.com/micro-ROS/micro_ros_arduino#patch-teensyduino).

Reviewer: ```suggestion
pre-compiled libraries as described [*in the micro_ros_arduino repository*](https://github.com/micro-ROS/micro_ros_arduino#patch-teensyduino).
```

and install by following this [*link for the Linux version*](https://www.arduino.cc/en/Guide/Linux) here.

After installing Arduino IDE download Teensyduino from this [*link here*](https://www.pjrc.com/teensy/td_154/TeensyduinoInstall.linux64) 
and follow the instruction as shown [*on this page*](https://www.pjrc.com/teensy/td_154/TeensyduinoInstall.linux64). 
To summarize the instructions which are as follow:

```
1. Download the Linux udev rules and copy the file to /etc/udev/rules.d.
https://www.pjrc.com/teensy/00-teensy.rules

2. type the following command in a terminal 
$ sudo cp 00-teensy.rules /etc/udev/rules.d/

3. Download and extract one of Arduino's Linux packages.
Note: Arduino from Linux distro packages is not supported.

4. Download the corresponding Teensyduino installer.

5. Run the installer in a termincal by adding execute permission and then execute it.
$ chmod 755 TeensyduinoInstall.linux64
$ ./TeensyduinoInstall.linux64
```
Now let's set up the patch for the teensy Arduino to use the pre-compiled micro-ros-client 
libraries, Open a terminal window and follow the commands below: For more information follow 

> the GitHub link from [*micro-ros-arduino*](https://github.com/micro-ROS/micro_ros_arduino/tree/foxy)

Reviewer: ```suggestion
the GitHub link from [*micro_ros_arduino*](https://github.com/micro-ROS/micro_ros_arduino/tree/foxy)
```

```bash
# for me it was $ export ARDUINO_PATH=/home/manzur/arduino-1.8.13/
export ARDUINO_PATH=[Your Arduino + Teensiduino path]

cd $ARDUINO_PATH/hardware/teensy/avr/

curl https://raw.githubusercontent.com/micro-ROS/micro_ros_arduino/foxy/extras/patching_boards/platform_teensy.txt > platform.txt
```

Once the above instruction is complete, we will now be able to use the Teensy 3.2 and 
program it with the pre-compiled micro-ros-client libraries using Arduino IDE.

> ## Program the Teensy: 

Reviewer: ```suggestion
## Program the Teensy
```

Now that we have patched the teensy Arduino IDE, we will be able to use the pre-compiled library by following these instructions:

1. Go to [*link to release section*](https://github.com/micro-ROS/micro_ros_arduino/releases) 
and download the last release of micro-ROS library for Arduino. 

> Place the file inside ```/home/$USERNAME/Arduino/libraries/ ``` as shown below.

Reviewer: ```suggestion
Place the file inside `/home/$USERNAME/Arduino/libraries/` as shown below.
```

![Illustration of patch location](patch_location.png)

Once this process is complete, now let us look at the example folder below:

![Illustration of example location](arduino_example_location.png)

For this tutorial and test, we will be using mico-ros-publisher example as shown above since this 
program will only publish integer data which will increase in every cycle. Once we selected the 

> example program, we will then upload the code in the teensy 3.2 connected to our host computer 

Reviewer: ```suggestion
example program, we will then upload the code in the Teensy 3.2 connected to our host computer 
```

which should show the result as follow.

![Illustration of upload completion](upload_completion.png)

> ## Running micro-ros-agent in ROS2 Foxy

Reviewer: ```suggestion
## Running micro-ROS agent in ROS 2 Foxy
```

> Now, let's disconnect the teensy for now from the host computer. We will then open a terminal 

Reviewer: ```suggestion
Now, let's disconnect the Teensy for now from the host computer. We will then open a terminal 
```

or in the docker run the agent program once again as shown at the end of 
step 2. Make sure to source the ROS path as below:

```bash

> #source /opt/ros/foxy/setup.bash

Reviewer: ```suggestion
source /opt/ros/foxy/setup.bash
```

> $ source /opt/ros/$ROS_DISTRO/setup.bash

Reviewer: ```suggestion
```
$ROS_DISTRO will be set in the setup.bash. Therefore, you have to specify the distro explicitly. As the whole tutorial is based on Foxy, it's fine to write `foxy` directly here.

```

and then run the agent program:

```bash
ros2 run micro_ros_agent micro_ros_agent serial --dev /dev/ttyACM0
```
Once the program is running it will show this message:

![Illustration of agent restart completion](agent_restart.png)

We will then reconnect the Teensy with the host computer and then we will see 
that the connection is complete and it shows like this:

![Illustration of agent connected](agent_connected.png)

This means the connection is complete with teensy containing micro-ros-client and micro-ros-agent in the host computer.
Now for the big moment and test the ROS topic published from the teensy. 
This time we will open another terminal or docker window and type as follow:

```bash
ros2 topic list
```
Which should list as shown below:

![Illustration of ros2 topic connected](ros2_topic_all.png)

> See, we have now ```/micro_ros_arduino_node_publisher``` topic publishing in the host computer. 

Reviewer: ```suggestion
See, we have now `/micro_ros_arduino_node_publisher` topic publishing in the host computer. 
```
